File: Code_Nicole/go/stocks/portfolio.go
package stocks
import "errors"

type Portfolio []Money

func (p Portfolio) Add(money Money) Money {
    p = append(p, money)
    return p
}

func (p Portfolio) Evaluate(currency string) Money {
    total := 0.0
    failedConversions := make([]string, 0)
    for _, m := range p {
        if convertedAmount, ok := convert(m, currency); ok {
            total = total + convertedAmount
        } else {
            failedConversions = append(failedConversions, m.currency + "->" + currency)
        }
    }

    if len(failedConversions) == 0 {
        return NewMoney(total, currency), nil
    }

    failures := "["
    for _, f := range failedConversions {
        failures = failures + f + ","
    }
    failures = failures + "]"

    return NewMoney(0, ""), errors.New("Missing exchange rate(s):" + failures)
}

func convert(money Money, currency string) (float64, bool) {
    exchangeRates := map[string]float64{
        "EUR->USD": 1.2,
        "USD->KRW": 1100.0,
    }
    // other rates: "USD->EUR": 1.0/1.2, "KRW->USD": 1.0/1100.0, "EUR->KRW": 1344.0, "KRW->EUR": 1.0/1344.0
    if money.currency == currency{
        return money.amount, true
    }
    key := money.currency + "=>" + currency
    rate, ok := exchangeRates[key]
    return money.amount * exchangeRates[key], ok
}

File: py/portfolio.py
from money import Money


class Portfolio:
    def __init__(self):
        self.moneys = []
        self._eur_to_usd = 1.2

    def add(self, *moneys):
        self.moneys.extend(moneys)

    def evaluate(self, bank, currency):
        # - using a lambda expression, we map the self.moneys array to a mpa of only the amounts in each Money object
        # - we then reduce this map to a single scalar value, using the operator.add operation
        # - we assign this scalar value to the variable named total
        # - we finally create a new Money object using this total and the currency passed in
        #   the first (and only) parameter to the evaluate method
        # - the last parameter to reduce (0 in our case) is the initial value of the accumulated result
        total = 0.0
        failures = []
        for m in self.moneys:
            try:
                total += bank.convert(m, currency).amount
            except Exception as ex:
                failures.append(ex)

        if len(failures) == 0:
            return Money(total, currency)

        failure_message = ",".join(f.args[0] for f in failures)  # str(f) has single-quote characters
        raise Exception("Missing exchange rate(s):[" + failure_message + "]")

